### ISSUE TITLE
[BUG] Character avatar pics not updating when changed.

### ISSUE THREAD

**Issue by Avroboros** [SillyTavern/SillyTavern#3199]
### Environment

🪟 Windows

### System

Firefox 133, Windows 11

### Version

1.12.9

### Desktop Information

>Node version: v22.11.0.
>Branch: Release

### Describe the problem

When you change a character's avatar, it doesn't update unless you close SillyTavern completely (F5 doesn't solve it).

Extremely annoying, especially when you need to modify the pics of multiple characters (and you need to check if they all look good or not).

### Please tick the boxes

- [X] I have explained the issue clearly, and I included all relevant info
- [X] I have checked that this [issue hasn't already been raised](https://github.com/SillyTavern/SillyTavern/issues?q=is%3Aissue)
- [X] I have checked the [docs](https://docs.sillytavern.app/) ![important](https://img.shields.io/badge/Important!-F6094E)

**Comment by cloak1505** [comment]
Was talked about several times in Discord, related to Firefox's aggressive image caching, presumably we can't fix it in a sane manner.

If I open Chrome, the new avatar shows up right away without closing ST WebServer.

**Comment by Avroboros** [comment]
> related to Firefox's aggressive image caching, presumably we can't fix it in a sane manner.

Got it, so is there a way to manually tamper with Firefox's image caching settings or is the only solution for now to use some other browser when running ST?

**Comment by Wolfsblvt** [comment]
> Got it, so is there a way to manually tamper with Firefox's image caching settings or is the only solution for now to use some other browser when running ST?

Yup. Most people will tell you to not use Firefox. So many issues regularly pop up for people using that browser with ST.

For the image caching itself, just hit CTRL+F5 to reload the page without cache. That should work.

**Comment by vorpal-void** [comment]
Another "workaround" is to add avatar image, then use Duplicate Character function. It creates second copy with updated avatar image info without page refresh.

**Comment by Avroboros** [comment]
Thanks for the suggestions. But I guess its not relevant anymore since I'm using Edge now. Firefox has become such a disappointment that even hearing its name pisses me off.

**Comment by language-coder** [comment]
It happened to me too. I discovered a way to solve this.
For someone that still uses firefox (such as myself).
It was quite simple to solve it, I went through by mere luck.
Just close the tab and open it again. The character now has the updated avatar.

**Comment by DeclineThyself** [comment]
@Cohee1207 May I modify image URLs to have the image file's date modified timestamp?
That should fix the caching issue. https://stackoverflow.com/a/321871

**Comment by Cohee1207** [comment]
Firefox users may try the [cache buster setting](https://docs.sillytavern.app/administration/config-yaml/#cache-buster-configuration) (it will force the browser to flush all caches after avatar update)

> May I modify image URLs to have the image file's date modified timestamp?

That is what people usually recommend, but I find this approach too naive. You'd basically have to add that logic everywhere where an avatar image can be rendered and make sure that after an update ever image is updated with a new query param, and that URLs with a timestamp are not saved to chat files.

**Comment by DeclineThyself** [comment]
Could [Cache-Control: no-cache, must-revalidate](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Cache-Control) work?

**Comment by Cohee1207** [comment]
With Firefox, you never know for sure. I tried a lot of possible solutions, and almost nothing worked. Even if the actual fetch response receives a new file, images on the page are _still_ going to be cached.

**Comment by DeclineThyself** [comment]
> Could [Cache-Control: no-cache, must-revalidate](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Cache-Control) work?

Fixed in https://github.com/SillyTavern/SillyTavern/pull/4743.
